glib/src/subclass/signal.rs

```diff
@@ -68,9 +69,117 @@ impl fmt::Debug for SignalInvocationHint {
     }
 }
 
+/// In-depth information of a specific signal
+#[repr(transparent)]
+pub struct SignalQuery(gobject_ffi::GSignalQuery);
+
+impl SignalQuery {
+    /// The name of the signal.
+    pub fn signal_name(&self) -> GString {
+        unsafe {
+            let name: GString = from_glib_none(self.0.signal_name);
+            name
+        }
+    }
+
+    /// The ID of the signal.
+    pub fn signal_id(&self) -> SignalId {
+        unsafe { SignalId::new(NonZeroU32::new_unchecked(self.0.signal_id)) }
+    }
+
+    /// The interface/instance this signal can be emitted for.
+    pub fn type_(&self) -> Type {
+        unsafe { from_glib(self.0.itype) }
+    }
+
+    /// The signal flags.
+    pub fn flags(&self) -> SignalFlags {
+        unsafe { from_glib(self.0.signal_flags) }
+    }
+
+    /// The return type for the user callback.
+    pub fn return_type(&self) -> Type {
+        unsafe { from_glib(self.0.return_type) }
+    }
+
+    /// The number of parameters the user callback takes.
+    pub fn n_params(&self) -> u32 {
+        self.0.n_params
+    }
+
+    /// The parameters for the user callback.
+    pub fn param_types(&self) -> Vec<Type> {
+        unsafe {
+            let types = self.0.param_types;
+            FromGlibContainerAsVec::from_glib_none_num_as_vec(types, self.n_params() as usize)
+        }
+    }
+}
+
+impl fmt::Debug for SignalQuery {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("SignalQuery")
+            .field("signal_name", &self.signal_name())
+            .field("type", &self.type_())
+            .field("flags", &self.flags())
+            .field("return_type", &self.return_type())
+            .field("param_types", &self.param_types())
+            .finish()
+    }
+}
 /// Signal ID.
 #[derive(Debug, Clone, Copy)]
-pub struct SignalId(pub(super) Type, pub(super) u32);
+pub struct SignalId(NonZeroU32);
+
+impl SignalId {
+    /// Create a new Signal Identifier.
+    ///
+    /// # Safety
+    ///
+    /// The caller has to ensure it's a valid signal identifier.
+    pub unsafe fn new(id: NonZeroU32) -> Self {
+        Self(id)
+    }
+
+    /// Find a SignalId by it's `name` and the `type` it connects to.
+    #[doc(alias = "g_signal_lookup")]
+    pub fn lookup(name: &str, type_: Type) -> Option<Self> {
+        unsafe {
+            let signal_id = gobject_ffi::g_signal_lookup(name.to_glib_none().0, type_.to_glib());
+            if signal_id == 0 {
+                None
+            } else {
+                Some(Self::new(NonZeroU32::new_unchecked(signal_id)))
+            }
+        }
+    }
+
+    pub fn id(&self) -> NonZeroU32 {
```

Comment:
As the type_ function is removed, I wonder if we can just impl Deref for u32/NonZeroU32? Not sure if it's really useful outside of unsafe ffi code.

Reply:
I wouldn't, but I would implement `FromGlib` and `ToGlib` instead

Reply:
Much better indeed.